lighthouse-core/computed/metrics/timing-summary.js

```diff
@@ -0,0 +1,106 @@
+const TraceOfTab = require('../trace-of-tab.js');
+const Speedline = require('../speedline.js');
+const FirstContentfulPaint = require('./first-contentful-paint.js');
+const FirstMeaningfulPaint = require('./first-meaningful-paint.js');
+const LargestContentfulPaint = require('./largest-contentful-paint.js');
+const FirstCPUIdle = require('./first-cpu-idle.js');
+const Interactive = require('./interactive.js');
+const SpeedIndex = require('./speed-index.js');
+const EstimatedInputLatency = require('./estimated-input-latency.js');
+const TotalBlockingTime = require('./total-blocking-time.js');
+
+class TimingSummary {
+  /**
+     * @param {LH.Artifacts} artifacts
+     * @param {LH.Audit.Context} context
+     * @return {Promise<LH.Artifacts.TimingSummary>}
+     */
+  static async summarize(artifacts, context) {
+    const trace = artifacts.traces[Audit.DEFAULT_PASS];
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const settings = context.settings;
+    const metricComputationData = {trace, devtoolsLog, settings};
+
+    /**
+     * @template TArtifacts
+     * @template TReturn
+     * @param {{request: (artifact: TArtifacts, context: LH.Audit.Context) => Promise<TReturn>}} Artifact
+     * @param {TArtifacts} artifact
+     * @return {Promise<TReturn|undefined>}
+     */
+    const requestOrUndefined = (Artifact, artifact) => {
+      return Artifact.request(artifact, context).catch(_ => undefined);
+    };
+
+    const traceOfTab = await TraceOfTab.request(trace, context);
+    const speedline = await Speedline.request(trace, context);
+    const firstContentfulPaint = await FirstContentfulPaint.request(metricComputationData, context);
+    const firstMeaningfulPaint = await FirstMeaningfulPaint.request(metricComputationData, context);
+    const largestContentfulPaint = await requestOrUndefined(LargestContentfulPaint, metricComputationData); // eslint-disable-line max-len
+    const firstCPUIdle = await requestOrUndefined(FirstCPUIdle, metricComputationData);
+    const interactive = await requestOrUndefined(Interactive, metricComputationData);
+    const speedIndex = await requestOrUndefined(SpeedIndex, metricComputationData);
+    const estimatedInputLatency = await EstimatedInputLatency.request(metricComputationData, context); // eslint-disable-line max-len
+    const totalBlockingTime = await TotalBlockingTime.request(metricComputationData, context); // eslint-disable-line max-len
+
+    /** @type {LH.Artifacts.TimingSummary} */
+    return {
+      // Include the simulated/observed performance metrics
+      firstContentfulPaint: firstContentfulPaint.timing,
+      firstContentfulPaintTs: firstContentfulPaint.timestamp,
+      firstMeaningfulPaint: firstMeaningfulPaint.timing,
+      firstMeaningfulPaintTs: firstMeaningfulPaint.timestamp,
+      // eslint-disable-next-line max-len
+      largestContentfulPaint: traceOfTab.lcpInvalidated ? undefined : largestContentfulPaint && largestContentfulPaint.timing,
```

Comment:
@connorjclark & @brendankenny you'll probably have opinions on this ^

The one awkward thing I encountered when moving this logic out of metrics.js was figuring out what to with the logic around lcpInvalidated. These code changes remove lcpInvalidated from the debug info, and instead lcp metrics aren't set if lcpInvalidated. You'd have a better perspective than me whether this is an okay change.

Reply:
oh, no, that's not what we want. `lcpInvalidated` is a debug flag signaling that the reason it was undefined was because it was invalidated. this would just lose that signal :)

maybe this artifact returns `{metrics, debugInfo}` or something where we can stick `lcpInvalidated`?

Reply:
Ah, that makes way more sense. I added a debug info property to the TimingSummary object.